dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/security/apikey/ApiTokenResolver.java

```diff
@@ -27,108 +27,156 @@
+        if ( validateParameterToken( request, parameterToken ) )
+            return hashToken( parameterToken );
+
+        return null;
     }
 
-    private String resolveFromAuthorizationHeader( HttpServletRequest request )
+    private char[] extractTokenFromHeader( HttpServletRequest request )
     {
         String authorization = request.getHeader( this.bearerTokenHeaderName );
         if ( !StringUtils.startsWithIgnoreCase( authorization, HEADER_TOKEN_KEY_PREFIX ) )
         {
-            return null;
+            return new char[0];
         }
 
-        Matcher matcher = authorizationPattern.matcher( authorization );
+        Matcher matcher = AUTHORIZATION_PATTERN.matcher( authorization );
         if ( !matcher.matches() )
         {
-            throw new ApiTokenAuthenticationException( ApiTokenErrors.invalidToken( "Api token is malformed" ) );
+            throw new ApiTokenAuthenticationException( ApiTokenErrors.invalidRequest( "Api token is malformed" ) );
+        }
+
+        return matcher.group( "token" ).toCharArray();
+    }
+
+    private static boolean validateHeaderToken( char[] authorizationHeaderToken, char[] parameterToken )
+    {
+        if ( authorizationHeaderToken.length > 0 )
+        {
+            checksum( authorizationHeaderToken );
+
+            if ( parameterToken.length > 0 )
+            {
+                throw new ApiTokenAuthenticationException( ApiTokenErrors
+                    .invalidRequest( "Found multiple tokens in the request" ) );
+            }
+            return true;
         }
 
-        return matcher.group( "token" );
+        return false;
     }
 
-    private static String resolveFromRequestParameters( HttpServletRequest request )
+    private static char[] extractTokenFromParameters( HttpServletRequest request )
     {
         String[] values = request.getParameterValues( REQUEST_PARAMETER_NAME );
         if ( values == null || values.length == 0 )
         {
-            return null;
+            return new char[0];
         }
 
         if ( values.length == 1 )
         {
-            return values[0];
+            return values[0].toCharArray();
         }
 
         throw new ApiTokenAuthenticationException(
             ApiTokenErrors.invalidRequest( "Found multiple Api tokens in the request" ) );
     }
 
-    private boolean isParameterTokenSupportedForRequest( HttpServletRequest request )
+    private boolean validateParameterToken( HttpServletRequest request, char[] parameterToken )
+    {
+        if ( parameterToken.length > 0 && isParameterTokenSupportedForRequest( request ) )
+        {
+            checksum( parameterToken );
+            return true;
+        }
+        return false;
+    }
+
+    private static void checksum( char[] token )
+    {
+        try
+        {
+            if ( !validateChecksum( token ) )
+            {
+                throw new ApiTokenAuthenticationException( ApiTokenErrors
```

Comment:
Always hard to say if exceptions are only used within the service layer but never meant to leave it. Just want to emphasis that if this is not caught and rethrown as one of our feedback exceptions we should use those here instead.